src/game/Tactical/Soldier_Control.cc

```diff
@@ -6364,90 +6364,29 @@ INT16 GetDirectionToGridNoFromGridNo( INT16 sGridNoDest, INT16 sGridNoSrc )
 }
 
 
-#if 0
-UINT8  atan8( INT16 x1, INT16 y1, INT16 x2, INT16 y2 )
-{
-	static int trig[8] = { 2, 3, 4, 5, 6, 7, 8, 1 };
-	// returned values are N=1, NE=2, E=3, SE=4, S=5, SW=6, W=7, NW=8
-	double dx=(x2-x1);
-	double dy=(y2-y1);
-	double a;
-	int i,k;
-	if (dx==0)
-		dx=0.00390625; // 1/256th
-#define PISLICES (8)
-	a=(atan2(dy,dx) + PI/PISLICES)/(PI/(PISLICES/2));
-	i=(int)a;
-	if (a>0)
-		k=i; else
-	if (a<0)
-		k=i+(PISLICES-1); else
-		k=0;
-	return(trig[k]);
-}
-#endif
-
-//#if 0
+// Returns the direction (in enum WorldDirections terms) from point 1 to point 2.
 UINT8 atan8( INT16 sXPos, INT16 sYPos, INT16 sXPos2, INT16 sYPos2 )
 {
-	DOUBLE test_x = sXPos2 - sXPos;
-	DOUBLE test_y = sYPos2 - sYPos;
-	UINT8  mFacing = WEST;
-	DOUBLE angle;
+	const double x = sXPos2 - sXPos;
 
-	if ( test_x == 0 )
-	{
-		test_x = 0.04;
-	}
+	// Negate y because in screen coordinates smaller y values are to the north
+	// but in the coordinate systen used by atan2 larger values are to the "north".
+	const double y = -(sYPos2 - sYPos);
 
-	angle = atan2( test_x, test_y );
-
-	do
-	{
-		if ( angle >=-PI*.375 && angle <= -PI*.125 )
-		{
-			mFacing = SOUTHWEST;
-			break;
-		}
+	// The result of atan2(0, 0) is implementation dependant, ensure we always
+	// return the same result for this case.
+	if (x == 0.0 && y == 0.0) return EAST;
 
-		if ( angle <= PI*.375 && angle >= PI*.125 )
-		{
-			mFacing = SOUTHEAST;
-			break;
-		}
+	double theta = std::atan2(y, x);  // theta now ∈ [-π, +π] = [-180°, 180°]
+	if (theta < 0) theta += 2 * M_PI; // theta now ∈ [0, +2π] = [0°, 360°]
```

Comment:
Isn't theta now in [π, 3π]? It still has the same range of values, just shifted, which I guess explains the extra M_PI below.

Reply:
No, if theta is negative it's in the range -π, 0. Add 2 * π to that and you arrive at π, 2π. Basically this line adds the part that was negative on top of the positive part.

Reply:
My comment was aimed at the comment, but I see now that it's referring to all cases, not just the negative one.